lib/Dialect/Synth/Transforms/CutRewriter.cpp

```diff
@@ -0,0 +1,564 @@
+  for (unsigned i = 0; i < cuts.size(); ++i) {
+    auto &cut = cuts[i];
+    // Create a unique identifier for the cut based on its inputs.
+    auto inputs = cut.inputs.getArrayRef();
+
+    // If the cut is a duplicate, skip it.
+    if (uniqueCuts.contains({inputs, cut.getRoot()}))
+      continue;
+
+    if (i != uniqueCount) {
+      // Move the unique cut to the front of the vector
+      // This maintains the order of cuts while removing duplicates
+      // by swapping with the last unique cut found.
+      cuts[uniqueCount] = std::move(cuts[i]);
+    }
+
+    // Beaware of lifetime of ArrayRef. `cuts[uniqueCount]` is always valid
+    // after this point.
+    uniqueCuts.insert(
+        {cuts[uniqueCount].inputs.getArrayRef(), cuts[uniqueCount].getRoot()});
+    ++uniqueCount;
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "Original cuts: " << cuts.size()
+                          << " Unique cuts: " << uniqueCount << "\n");
+  // Resize the cuts vector to the number of unique cuts found
+  cuts.resize(uniqueCount);
+
+  // Maintain size limit by removing worst cuts
+  if (cuts.size() > options.maxCutSizePerRoot) {
+    // Sort by priority using heuristic. Currently we sort by (depth, input
+    // size).
+    // TODO: Make this configurable.
+    // TODO: Implement pruning based on dominance.
+
+    std::sort(cuts.begin(), cuts.end(), [](const Cut &a, const Cut &b) {
+      if (a.getDepth() == b.getDepth())
+        return a.getInputSize() < b.getInputSize();
+      return a.getDepth() < b.getDepth();
+    });
+
+    cuts.resize(options.maxCutSizePerRoot);
+  }
+
+  isFrozen = true; // Mark the cut set as frozen
+}
+
+//===----------------------------------------------------------------------===//
+// CutEnumerator
+//===----------------------------------------------------------------------===//
+
+CutEnumerator::CutEnumerator(const CutRewriterOptions &options)
+    : options(options) {}
+
+CutSet *CutEnumerator::createNewCutSet(Value value) {
+  auto [cutSetPtr, inserted] =
+      cutSets.try_emplace(value, std::make_unique<CutSet>());
+  assert(inserted && "Cut set already exists for this value");
+  return cutSetPtr->second.get();
+}
+
+llvm::MapVector<Value, std::unique_ptr<CutSet>> CutEnumerator::takeVector() {
+  return std::move(cutSets);
+}
+
+void CutEnumerator::clear() { cutSets.clear(); }
+
+LogicalResult CutEnumerator::visit(Operation *op) {
+  if (isSupportedLogicOp(op)) {
+    // Support only binary operations with a single bit
+    // TODO: This can be extended to support variadic multiplication.
+    if (op->getNumOperands() > 2)
+      return op->emitError() << "Cut enumeration supports at most 2 operands, "
+                             << "found: " << op->getNumOperands();
+    if (!op->getResult(0).getType().isInteger(1))
+      return op->emitError()
+             << "Supported logic operations must have a single bit "
+                "result type but found: "
+             << op->getResult(0).getType();
```

Comment:
It looks like these checks are performed in `visitLogicOp` again. Might be worth removing these ones here in favor of `visitLogicOp`?